### PR TITLE
Homework 8  JDI lesson 1

### PR DESCRIPTION
**Pull request by sciencefactor**

### CODE REVIEW

File: src/main/java/com/epam/tc/gdi/hw1/site/entities/MetalColorEntity.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package com.epam.tc.gdi.hw1.site.entities;
+
+import java.util.List;
+
+public class MetalColorEntity {
```
Comment (pnatashap) [inline src/main/java/com/epam/tc/gdi/hw1/site/entities/MetalColorEntity.java]: Should extend DataClass

File: src/main/java/com/epam/tc/gdi/hw1/site/entities/User.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package com.epam.tc.gdi.hw1.site.entities;
+
+public class User  {
```
Comment (pnatashap) [inline src/main/java/com/epam/tc/gdi/hw1/site/entities/User.java]: Should extend DataClass

File: src/main/java/com/epam/tc/gdi/hw1/site/pages/MetalsColorsPage.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+public class MetalsColorsPage extends WebPage {
+
+    @Css("form[class=form]")
+    public MetalsColorsForm metalsColorsForm;
+
+    @Css("ul[class=\"panel-body-list results\"]>li")
+    public WebList result;
+
+    public void shouldBeOpened() {
+        if (!getTitle().equalsIgnoreCase("metals and colors")) {
```
Comment (pnatashap) [inline src/main/java/com/epam/tc/gdi/hw1/site/pages/MetalsColorsPage.java]: The better way is to create annotation Title and use built-in method to check title

File: src/main/java/com/epam/tc/gdi/hw1/site/pages/components/complex/MetalsColorsForm.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    public Dropdown metals;
+
+    @FindBy(id = "salad-dropdown")
+    public CheckSelector vegetables;
+
+    @FindBy(id = "submit-button")
+    public Button submit;
+
+    @JDIAction("Fill '{name}' with {0}")
+    public void fill(MetalColorEntity entity) {
```
Comment (pnatashap) [inline src/main/java/com/epam/tc/gdi/hw1/site/pages/components/complex/MetalsColorsForm.java:43]: + @Override

File: src/main/java/com/epam/tc/gdi/hw1/site/pages/components/complex/MetalsColorsForm.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        metals.select(entity.metals);
+        vegetables.check(entity.vegetables.toArray(String[]::new));
+        elements.checklist.check(entity.elements.toArray(String[]::new));
+    }
+
+    @JDIAction(
+        value = "Submit '{name}' with {0}",
+        timeout = 0
+    )
+    public void submit(MetalColorEntity entity) {
```
Comment (pnatashap) [inline src/main/java/com/epam/tc/gdi/hw1/site/pages/components/complex/MetalsColorsForm.java:58]: +@Override

File: src/main/java/com/epam/tc/gdi/hw1/site/pages/components/simple/HeaderNavigation.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import com.epam.jdi.light.elements.composite.Section;
+import com.epam.jdi.light.elements.pageobjects.annotations.FindBy;
+import com.epam.jdi.light.elements.pageobjects.annotations.locators.Css;
+import com.epam.jdi.light.ui.html.elements.common.Button;
+import com.epam.jdi.light.ui.html.elements.common.Link;
+
+
+public class HeaderNavigation extends Section {
+
+    @FindBy(css = "a[href=\"index.html\"]")
```
Comment (pnatashap) [inline src/main/java/com/epam/tc/gdi/hw1/site/pages/components/simple/HeaderNavigation.java]: It's easier to use @FindBy(css = "a[href='index.html']") and do not escape symbols (and easy to understand)

File: src/test/java/com/epam/tc/gdi/hw1/tests/TestsInit.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    @BeforeMethod
+    void loginIfNot() {
+        Preconditions.loginIfNot();
+    }
+
+    @AfterTest
+    void tearDown() {
+        WebDriverUtils.killAllSeleniumDrivers();
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/gdi/hw1/tests/TestsInit.java:26]: It's not required to kill browser after each test, it's enough to close all after suite

File: src/test/java/com/epam/tc/hw2/TestControlElementsOfElementsPage.java [hunk 8]
```diff
@@ -1,14 +1,12 @@
 package com.epam.tc.hw2;
 
-import io.github.bonigarcia.wdm.WebDriverManager;
 import java.util.HashMap;
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw2/TestControlElementsOfElementsPage.java]: this change should not be a part of this PR

File: src/test/java/com/epam/tc/hw3/tests/TestElementsPage.java [hunk 9]
```diff
@@ -1,11 +1,8 @@
 package com.epam.tc.hw3.tests;
 
 import com.epam.tc.hw3.pages.DifferentElementsPageObject;
-import com.epam.tc.hw3.pages.HomePageObject;
 import com.epam.tc.hw3.utils.PagesDataProviders;
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw3/tests/TestElementsPage.java]: this change should not be a part of this PR

File: src/test/java/com/epam/tc/hw3/tests/TestHomePageControlElements.java [hunk 10]
```diff
@@ -1,11 +1,8 @@
 package com.epam.tc.hw3.tests;
 
 
-import com.epam.tc.hw3.pages.HomePageObject;
 import com.epam.tc.hw3.utils.PagesDataProviders;
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw3/tests/TestHomePageControlElements.java]: this change should not be a part of this PR

**Review by pnatashap — CHANGES_REQUESTED** [review]
Comments are in a code

File: src/test/java/com/epam/tc/hw3/tests/TestHomePageControlElements.java (hunk 10 above)
Comment (sciencefactor) [reply src/test/java/com/epam/tc/hw3/tests/TestHomePageControlElements.java]: Done

File: src/test/java/com/epam/tc/hw3/tests/TestElementsPage.java (hunk 9 above)
Comment (sciencefactor) [reply src/test/java/com/epam/tc/hw3/tests/TestElementsPage.java]: Done

File: src/test/java/com/epam/tc/hw2/TestControlElementsOfElementsPage.java (hunk 8 above)
Comment (sciencefactor) [reply src/test/java/com/epam/tc/hw2/TestControlElementsOfElementsPage.java]: Done

File: src/test/java/com/epam/tc/gdi/hw1/tests/TestsInit.java (hunk 7 above)
Comment (sciencefactor) [reply src/test/java/com/epam/tc/gdi/hw1/tests/TestsInit.java:26]: Done

File: src/main/java/com/epam/tc/gdi/hw1/site/pages/components/simple/HeaderNavigation.java (hunk 6 above)
Comment (sciencefactor) [reply src/main/java/com/epam/tc/gdi/hw1/site/pages/components/simple/HeaderNavigation.java]: Done

File: src/main/java/com/epam/tc/gdi/hw1/site/pages/components/complex/MetalsColorsForm.java (hunk 5 above)
Comment (sciencefactor) [reply src/main/java/com/epam/tc/gdi/hw1/site/pages/components/complex/MetalsColorsForm.java:58]: Done

File: src/main/java/com/epam/tc/gdi/hw1/site/pages/components/complex/MetalsColorsForm.java (hunk 4 above)
Comment (sciencefactor) [reply src/main/java/com/epam/tc/gdi/hw1/site/pages/components/complex/MetalsColorsForm.java:43]: Done

File: src/main/java/com/epam/tc/gdi/hw1/site/entities/User.java (hunk 2 above)
Comment (sciencefactor) [reply src/main/java/com/epam/tc/gdi/hw1/site/entities/User.java]: Done

File: src/main/java/com/epam/tc/gdi/hw1/site/entities/MetalColorEntity.java (hunk 1 above)
Comment (sciencefactor) [reply src/main/java/com/epam/tc/gdi/hw1/site/entities/MetalColorEntity.java]: Done

File: src/main/java/com/epam/tc/gdi/hw1/site/pages/MetalsColorsPage.java (hunk 3 above)
Comment (sciencefactor) [reply src/main/java/com/epam/tc/gdi/hw1/site/pages/MetalsColorsPage.java]: Done

File: src/main/java/com/epam/tc/gdi/hw1/site/pages/components/simple/HiddenCheckList.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+import com.epam.jdi.light.common.JDIAction;
+import com.epam.jdi.light.elements.complex.Checklist;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+
+public class HiddenCheckList extends Checklist {
```
Comment (pnatashap) [inline src/main/java/com/epam/tc/gdi/hw1/site/pages/components/simple/HiddenCheckList.java:11]: Checklist can be used if use locator to input

File: src/test/java/com/epam/tc/gdi/hw1/tests/steps/Preconditions.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
+package com.epam.tc.gdi.hw1.tests.steps;
+
+import com.epam.tc.gdi.hw1.site.entities.Users;
+import com.epam.tc.gdi.hw1.site.pages.components.complex.Header;
+
+public class Preconditions {
+
+    public static void loginIfNot() {
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/gdi/hw1/tests/steps/Preconditions.java]: It's better to use the same notation: shouldBeLoggedIn

**Review by pnatashap — APPROVED** [review]
Good

File: src/main/java/com/epam/tc/gdi/hw1/site/pages/components/simple/HiddenCheckList.java (hunk 11 above)
Comment (sciencefactor) [reply src/main/java/com/epam/tc/gdi/hw1/site/pages/components/simple/HiddenCheckList.java:11]: Yes, [type="checkbox"] is input elements.

File: src/test/java/com/epam/tc/gdi/hw1/tests/steps/Preconditions.java (hunk 12 above)
Comment (sciencefactor) [reply src/test/java/com/epam/tc/gdi/hw1/tests/steps/Preconditions.java]: Renamed

**Pull request merged**